processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -231,15 +238,17 @@ public int lookupId(String name)
   private final Map<String, Integer> metricIndexes;
   private final Map<String, String> metricTypes;
   private final ImmutableList<String> metricNames;
-  private final LinkedHashMap<String, Integer> dimensionOrder;
+  private final ConcurrentMap<String, Integer> dimensionOrder = new ConcurrentHashMap<>();
+  private final AtomicInteger orderIncrementer = new AtomicInteger(0);
+  private final AtomicInteger rowIncrementer = new AtomicInteger(0);
+  private static final Integer raceChecker = -1; // Used in settling race elections
   private final AggregatorType[] aggs;
   private final DimensionHolder dimValues;
-  private final Map<String, ColumnCapabilitiesImpl> columnCapabilities;
+  private final ConcurrentMap<String, ColumnCapabilitiesImpl> columnCapabilities = new ConcurrentHashMap<>();
   private final boolean deserializeComplexMetrics;
 
-  protected final CopyOnWriteArrayList<String> dimensions;
 
-  private volatile AtomicInteger numEntries = new AtomicInteger();
+  protected volatile AtomicInteger numEntries = new AtomicInteger();
```

Comment:
It was actually part of my initial debugging as I needed it in OffHeap and OnHeap subclasses. But not anymore